### DISCUSSION TITLE
TabControl - Show TabItem in just one line

### DISCUSSION THREAD

**Discussion by marcosgerene** [AvaloniaUI/Avalonia#11896] [Q&A] [answered]
Hi everyone,

I am pretty new in Avalonia, I don't know it here is the best place to ask those things, so please let me know if I am in the wrong place.

To be quick, what I want to do is this:
https://stackoverflow.com/questions/26764401/make-tabcontrol-headers-scrollable-in-wpf

I have done this in WPF before, but it does not work with Avalonia. How can I have the same behavior?

**Comment by maxkatz6** [comment]
With build-in Avalonia styles it is in one line by default.

**Reply by marcosgerene** [reply]
@maxkatz6 It is going to the next line when does not fit in the line.

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/668f999a-bf9d-4dab-8b6e-9114095bb473)

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/ab53559f-6d3e-42af-badf-ae46977977f1)

Example code:

```
<TabControl>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="LastOne">
		<Grid />
	</TabItem>
</TabControl>
```

**Reply by timunie** [reply]
Which Avalonia version do you use?

**Reply by marcosgerene** [reply]
![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/96a45e21-dd2d-48b8-bf43-5fa424b6f869)

11.0.0-rc1.1

**Comment by rabbitism** [comment]
It is quite similar, you need to rewrite control template. If you are using Fluent theme, you can check source code.

https://github.com/AvaloniaUI/Avalonia/blob/e0aa8609a75254bd2c28ab2224120e5e3ef9dd97/src/Avalonia.Themes.Fluent/Controls/TabControl.xaml#L35-L37

Here you can wrap ItemsPresenter into a scrollviewer, and set ItemsPanel as horizontal StackPanel.

**Reply by marcosgerene** [reply]
@rabbitism  Thanks for your help.

My new code is:
```
<TabControl>
	<TabControl.Styles>
		<Style Selector="TabControl">
			<Setter Property="Template">
				<ControlTemplate>
					<Grid>
						<ScrollViewer HorizontalScrollBarVisibility="Auto">
							<ItemsPresenter Name="PART_ItemsPresenter"
											ItemsPanel="{TemplateBinding ItemsPanel}"
											DockPanel.Dock="{TemplateBinding TabStripPlacement}" />
						</ScrollViewer>
					</Grid>
				</ControlTemplate>
			</Setter>
		</Style>
	</TabControl.Styles>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="TabItem">
		<Grid />
	</TabItem>
	<TabItem Header="LastOne">
		<Grid />
	</TabItem>
</TabControl>
```

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/66f40596-37b9-4103-b94d-9a23a81292cf)

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/0777b3d9-cb92-4ed7-9360-0daa16e0111c)


The problem here is the scroll position. I want to scroll between the tabs as WPF did on this:

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/ca7d27fb-1e8f-4210-a091-4b702ce07844)

In that example I sent I am scrolling the full content.

**Reply by marcosgerene** [reply]
There is another problem when I changed that:

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/4b4e1656-2361-4382-9cbc-b7df8b592067)

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/d650fd44-fce5-486d-bbf3-6355dd51a91e)

My textblock disappeared.

**Comment by timunie** [comment]
Modify this in App.Resources: https://github.com/AvaloniaUI/Avalonia/blob/e0aa8609a75254bd2c28ab2224120e5e3ef9dd97/src/Avalonia.Themes.Fluent/Controls/TabStrip.xaml#L14-L32

Should be similar to this:

```diff
  <ControlTheme x:Key="{x:Type TabStrip}" TargetType="TabStrip">
    <Setter Property="Template">
      <ControlTemplate>
        <Border Background="{TemplateBinding Background}"
                BorderBrush="{TemplateBinding BorderBrush}"
                BorderThickness="{TemplateBinding BorderThickness}"
                CornerRadius="{TemplateBinding CornerRadius}"
                Padding="{TemplateBinding Padding}">
+         <ScrollViewer>
             <ItemsPresenter Name="PART_ItemsPresenter"
                          ItemsPanel="{TemplateBinding ItemsPanel}" />
+          </ScrollViewer>
         </Border>
      </ControlTemplate>
    </Setter>
    <Setter Property="ItemsPanel">
      <ItemsPanelTemplate>
!        <StackPanel />
      </ItemsPanelTemplate>
    </Setter>
  </ControlTheme>
````

**Reply by marcosgerene** [reply]
Did I forget something?

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/73ac081a-54c3-4f60-84a9-7219e63dfb7e)

**Reply by timunie** [reply, marked as answer]
🙈 my bad. You need to adjust `TabControl`-Template instead in a similar way. 
https://github.com/AvaloniaUI/Avalonia/blob/e0aa8609a75254bd2c28ab2224120e5e3ef9dd97/src/Avalonia.Themes.Fluent/Controls/TabControl.xaml

```xml
<ControlTheme x:Key="{x:Type TabControl}" TargetType="TabControl">
    <Setter Property="Template">
	    <ControlTemplate>
		    <Border BorderBrush="{TemplateBinding BorderBrush}"
		            BorderThickness="{TemplateBinding BorderThickness}"
		            CornerRadius="{TemplateBinding CornerRadius}"
		            Background="{TemplateBinding Background}"
		            HorizontalAlignment="{TemplateBinding HorizontalAlignment}"
		            VerticalAlignment="{TemplateBinding VerticalAlignment}">
			    <DockPanel>
				    <ScrollViewer DockPanel.Dock="{TemplateBinding TabStripPlacement}" 
				                  HorizontalAlignment="Stretch"
				                  HorizontalScrollBarVisibility="Auto">
					    <ItemsPresenter Name="PART_ItemsPresenter"
					                    ItemsPanel="{TemplateBinding ItemsPanel}" />
				    </ScrollViewer >
				    <ContentPresenter Name="PART_SelectedContentHost"
				                      Margin="{TemplateBinding Padding}"
				                      HorizontalContentAlignment="{TemplateBinding HorizontalContentAlignment}"
				                      VerticalContentAlignment="{TemplateBinding VerticalContentAlignment}"
				                      Content="{TemplateBinding SelectedContent}"
				                      ContentTemplate="{TemplateBinding SelectedContentTemplate}" />
			    </DockPanel>
		    </Border>
	    </ControlTemplate>
    </Setter>
    <Setter Property="ItemsPanel">
	    <ItemsPanelTemplate>
		           <StackPanel Orientation="Horizontal"/>
	    </ItemsPanelTemplate>
    </Setter>
</ControlTheme>
```

**Reply by marcosgerene** [reply]
@timunie 

Thank you very much, it works now!

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/864a7c34-d6ed-4003-893c-a73de5d95bbc)

![image](https://github.com/AvaloniaUI/Avalonia/assets/15945003/02d0ec33-0dbe-4a43-8829-c307080fef0f)
